README.md
## Gputweak Script

This is a gputweak helper script using nvidia-smi, nvidia-settings and rocm-smi.  
Script will be able to find GPU automatically and use the define tweak inside config/gputweak.json base on GPU product and vendor ID.  

## Requirement

- Ubuntu 16.04 or above
- jq library installed ( apt-get install -y jq )
- Install either official NVIDIA driver or AMDGPU driver, FGLRX is not supported.

## Install

This will install script in /opt/gputweak and enable startup script
Note that /etc/systemd/system/default.target will be change to script custom.target
/etc/X11/xorg.conf will be created for dummy monitor and locked using chattr 


```
sudo ./install.sh
```

## Usage

```
Usage: gputweak [-v|--verbose] [-h|--help] [OPTIONS]
 -v | --verbose   : output verbose
 -h | --help      : this message

 OPTIONS
  --test           : Test mode, no tweak is run. Using
                      NVIDIA  10de:1b81 10de:119d device
                      AMD     1002:67df 1462:3417 device
  --coolbits       : set NVIDIA coolbits, default is 12
  --nvsmi          : set nvidia-smi executable path, default /usr/bin/nvidia-smi
  --nvset          : set nvidia-settings executable path, default /usr/bin/nvidia-settings
  --rocmsmi        : set rocm-smi executable path, default /usr/bin/rocm-smi
  --reset          : reset and remove all tweak, will ignore other options
  --config         : specify gputweak config file, overwrite default. check config/gputweak.json for example
  --user           : read gputweak config file from /home/{user}/.gputweak/gputweak.json
```

## Run at startup
Copy all file and folder inside config/systemd to /etc/systemd/system. 
If you run install.sh above, you can skip this.

```bash
sudo cp -r config/systemd/* config/systemd/system
```

Set custom.target as default target then reboot

```bash
sudo rm -f /etc/systemd/system/default.target
sudo ln -s /etc/systemd/system/custom.target /etc/systemd/system/default.target
```

## Create GPU tweak config

You can add more tweak config in config/gputweak.json. Below are the step

1. Find your GPU ID check from your console using "lspci -vnn", then take note of the product and vendor id for both system and subsytem one.  
![pci_id](https://raw.githubusercontent.com/wiryonolau/gputweak/master/img/pci_id.jpg)

2. Then add another json object inside the config with this format
```
[
    {
        "description": "Put the description of your GPU here" (optional),
        "driver" : "specify the driver type, either amdgpu or nvidia, the script have to use this value (required)",
        "system_id": "see image above on how to obtain, if not available use 0000:0000 (required)",
        "subsystem_id": "see image above on how to obtain, if not available use 0000:0000 (required)",
        "tweaks" : [
            "a list of tweak you want to run (required)"
        ]
    }
]
```

Example tweak for AMDGPU, setting fan and core clock
```
[
    {
        "system_id" : "..."
        "subsystem_id" : "..."
        "tweaks" : [      
            "@ROCSMI@ -d @GPUID@ --setfan 180",
            "@ROCSMI@ -d @GPUID@ --setsclk 3"
        ]
    }
]
```

Example tweak for NVIDIA
```
[
    {
        "system_id" : "..."
        "subsystem_id" : "..."
        "tweaks" : [
            "@NVSET@ -a [gpu:0]/GPUFanControlState=1",
            "@NVSET@ -a [fan:0]/GPUTargetFanSpeed=70",
            "@NVSET@ -a [gpu:0]/GPUPowerMizerMode=1",
            "@NVSET@ -a [gpu:0]/GPUGraphicsClockOffset[3]=-200",
            "@NVSET@ -a [gpu:0]/GPUMemoryTransferRateOffset[3]=450",
            "@NVSMI@ -i @PCI_BUS@ -pm 1",
            "@NVSMI@ -i @PCI_BUS@ -pl 100"
        ]
    }
]
```

Predefine variable for tweaks are
- @NVSMI@ : will be replace by nvidia-smi executable path 
- @NVSET@ : will be replace by nvidia-settings executable path
- @PCI_BUS@ : will be replace by GPU pci bus address ( for NVIDIA card )
- @ROCSMI@ : will be replace by rocm-smi executable path
- @GPUID@ : will be replace by GPU id base on pci bus ( for AMD card )

3. Run the script directly

```
gputweak -v
```

## Notes
There is no safety mechanism when deploying tweak, so it's best to check the maximum clock, memclock, fan, etc before run this automatically.  
For AMD card currently there is no overclock mechanism available in Linux, you still require to update the BIOS directly.  
Script can be run manually if you need to reset or apply or test new setting.

## Disclaimer
Use this script at your own risk. We're not responsible for any damage this script cause to your system / GPU.

## Donate
Donation are welcome 

BTC - bc1qur8aeernt2s7982sffyz7yv6882vgewsqdkyjg ( Segwit )  
BTC - 3Kcs5hWVQRUGtJKv3DjKbkzbGTEtFEQkUx ( Compatible )  
ETH - 0x9beb9B182a273Da689b08228385137440fEb8D6B  


CHANGELOG.md
## 0.1.0 (August 21, 2019)
  - Add gputweak done state
  - disable run opencl for amd, cause setsclk to reset
  - add delay before applying opencl
  - add nocustom target
  - Fix install script
  - add noxorg options
  - Add user base tweak
  - Update config
  - Add function to update config from git
  - Fix typo condition
  - Add small opencl program to run on tweak to reduce watt
  - Readjust readme content
  - Fix readme tutorial
  - Update readme
  - Fix systemctl
  - Fix tweak amd
  - remove /bin/bash when running amd tweak, fix systemd script
  - Add 10s start delay to gputweak.service
  - fix typo DISPLAY
  - fix nvidia tweak
  - Add P104-100
  - Fix nvidia tweak, remove noxorg option
  - install main
  - fix typo and install script
  - fix amd tweak and install script
  - Add install jq library in install.sh
  - fix README
  - Fix install.sh, add xorg.conf
  - Add install script
  - add systemd config, add x detection
  - fix reset function, add new gpu
  - backup works
  - backup file
  - first commit



install.sh
#!/usr/bin/env bash
VERBOSE=""
NOUPDATE=""
NOXORG=""
NOCUSTOM=""

function usage
{
    echo "Usage: install.sh [-v|--verbose] [-h|--help] [--noupdate]"
    echo " -v | --verbose   : output verbose"
    echo " -h | --help      : Print this help"
    echo " --noupdate       : Don't run apt-get update"
    echo " --noxorg         : Don't replace xorg"
    echo " --nocustom       : Don't use custom target"
}

function parse_args
{
    local _ARGS=()

    while [ "$1" ]
    do
        case "$1" in
            "-v" | "--verbose" )
                VERBOSE=1
            ;;
            "-h" | "--help")
                usage
                exit 0
                ;;
            "--noupdate" )
                NOUPDATE=1
            ;;
            "--noxorg" )
                NOXORG=1
            ;;
            "--nocustom" )
                NOCUSTOM=1
            ;;
        esac
        shift
    done
    
    set -- "${args[@]}"

    start_install
    exit 0
}

function start_install
{
    if [[ $EUID -ne 0 ]]; then
       printf "This script must be run as root\n" 
       exit 1
    fi

    #Install jq (bash json parser) library
    if [ -z "${NOUPDATE}" ]; then
        apt-get update
    fi
    apt install -y jq

    #Install opencl-headers for testing
    apt install -y opencl-headers

    #Put script in opt
    printf "Create gputweak executable\n"
    WORKDIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
    mkdir -p /opt/gputweak
    cp -r ${WORKDIR}/* /opt/gputweak
    rm -rf /usr/bin/gputweak
    ln -s /opt/gputweak/gputweak /usr/bin/gputweak

    #Build AMDGPU OPENCL device detector
    if [[ -z "/opt/gputweak/amdgpu-devices.c" ]]; then
        /usr/bin/wget https://laanwj.github.io/assets/2016/05/06/opencl-ubuntu1604/devices.c -O /opt/gputweak/amdgpu-devices.c 
    fi
    /usr/bin/gcc /opt/gputweak/amdgpu-devices.c -o /opt/gputweak/amdgpu-devices -O2 /opt/amdgpu-pro/lib/x86_64-linux-gnu/libOpenCL.so

    #Setup main xorg.conf
    if [ -z "${NOXORG}" ]; then
        printf "Setup xorg.conf\n"
        if [[ ! -z "/etc/X11/xorg.conf" ]]; then
            NOW=$(date +%s)
            chattr -i /etc/X11/xorg.conf
            cp /etc/X11/xorg.conf "/etc/X11/xorg.conf.${NOW}"
        fi
        cp /opt/gputweak/config/xorg.conf /etc/X11/xorg.conf
        chattr +i /etc/X11/xorg.conf
    fi

    #Enable systemd custom target startup
    printf "Setup systemd startup script\n"
    systemctl disable gputweak.service
    cp ${WORKDIR}/config/systemd/* /etc/systemd/system
    if [ -z "${NOCUSTOM}" ]; then
        rm -f /etc/systemd/system/default.target
        ln -s /etc/systemd/system/custom.target /etc/systemd/system/default.target
    fi
    systemctl enable gputweak.service

    systemctl daemon-reload
    printf "Setup done\n"
}

function main
{
    parse_args "$@"
    exit 0
}

main "$@"



amdgpu-devices.c
// Copyright (c) 2016 Wladimir J. van der Laan
// Distributed under the MIT software license.
// Based on an example from the OpenCL cookbook.

#include <stdio.h>
#include <stdlib.h>
#ifdef __APPLE__
#include <OpenCL/opencl.h>
#else
#include <CL/cl.h>
#endif

struct platform_data_item {
    int id;
    char *name;
};

struct platform_data_item platform_data_items[] = {
    { CL_PLATFORM_PROFILE, "Profile"},
    { CL_PLATFORM_VERSION, "Version"},
    { CL_PLATFORM_NAME,    "Name"},
    { CL_PLATFORM_VENDOR,  "Vendor"},
    { CL_PLATFORM_EXTENSIONS, "Extensions"},
};

#define ARRAYLEN(array)     (sizeof(array)/sizeof((array)[0]))

int main() {
 
    int i, j;
    char* value;
    size_t valueSize;
    cl_uint platformCount = 0;
    cl_platform_id* platforms;
    cl_uint deviceCount;
    cl_device_id* devices;
    cl_uint maxComputeUnits;
 
    // get all platforms
    if (clGetPlatformIDs(0, NULL, &platformCount) != CL_SUCCESS) {
        printf("Unable to get platform IDs\n");
        exit(1);
    }
    platforms = (cl_platform_id*) malloc(sizeof(cl_platform_id) * platformCount);
    if (clGetPlatformIDs(platformCount, platforms, NULL) != CL_SUCCESS) {
        printf("Unable to get platform IDs\n");
        exit(1);
    }
 
    for (i = 0; i < platformCount; i++) {
        printf("%i. Platform\n", i+1);

        char data[1024];
        size_t retsize;
        for (int j=0; j<ARRAYLEN(platform_data_items); ++j) {
            if (clGetPlatformInfo(platforms[i], platform_data_items[j].id, sizeof(data), data, &retsize) != CL_SUCCESS || retsize == sizeof(data)) {
                printf("Unable to get platform %s\n", platform_data_items[j].name);
                continue;
            }
            printf("  %s: %s\n", platform_data_items[j].name, data);
        }
 
        // get all devices
        if (clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_ALL, 0, NULL, &deviceCount) != CL_SUCCESS) {
            printf("Unable to get device IDs for platform %p\n", platforms[i]);
            continue;
        }
        devices = (cl_device_id*) malloc(sizeof(cl_device_id) * deviceCount);
        if (clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_ALL, deviceCount, devices, NULL) != CL_SUCCESS) {
            printf("Unable to get device IDs for platform %p\n", platforms[i]);
            continue;
        }
 
        // for each device print critical attributes
        for (j = 0; j < deviceCount; j++) {
 
            // print device name
            clGetDeviceInfo(devices[j], CL_DEVICE_NAME, 0, NULL, &valueSize);
            value = (char*) malloc(valueSize);
            clGetDeviceInfo(devices[j], CL_DEVICE_NAME, valueSize, value, NULL);
            printf("%d. Device: %s\n", j+1, value);
            free(value);
 
            // print hardware device version
            clGetDeviceInfo(devices[j], CL_DEVICE_VERSION, 0, NULL, &valueSize);
            value = (char*) malloc(valueSize);
            clGetDeviceInfo(devices[j], CL_DEVICE_VERSION, valueSize, value, NULL);
            printf(" %d.%d Hardware version: %s\n", j+1, 1, value);
            free(value);
 
            // print software driver version
            clGetDeviceInfo(devices[j], CL_DRIVER_VERSION, 0, NULL, &valueSize);
            value = (char*) malloc(valueSize);
            clGetDeviceInfo(devices[j], CL_DRIVER_VERSION, valueSize, value, NULL);
            printf(" %d.%d Software version: %s\n", j+1, 2, value);
            free(value);
 
            // print c version supported by compiler for device
            clGetDeviceInfo(devices[j], CL_DEVICE_OPENCL_C_VERSION, 0, NULL, &valueSize);
            value = (char*) malloc(valueSize);
            clGetDeviceInfo(devices[j], CL_DEVICE_OPENCL_C_VERSION, valueSize, value, NULL);
            printf(" %d.%d OpenCL C version: %s\n", j+1, 3, value);
            free(value);
 
            // print parallel compute units
            clGetDeviceInfo(devices[j], CL_DEVICE_MAX_COMPUTE_UNITS,
                    sizeof(maxComputeUnits), &maxComputeUnits, NULL);
            printf(" %d.%d Parallel compute units: %d\n", j+1, 4, maxComputeUnits);
 
        }
 
        free(devices);
 
    }
 
    free(platforms);
    return 0;
 
}


tweak.sh
#!/bin/bash

#Global Variables
DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
TWEAKCONFIG="$DIR/config/gputweak.json"
NVSMI='/usr/bin/nvidia-smi'
NVSET='/usr/bin/nvidia-settings'
ROCSMI='/usr/bin/rocm-smi'
TEST=0
VERBOSE=0
DISPLAY=10
PCI_REGEX='/\(VGA compatible\|3D controller\).*\(AMD\|\NVIDIA\)/s/^\(..\):\(..\).\(.\).*/0x\1:0x\2:0x\3/p'
SYSID_REGEX='\s(VGA compatible|3D controller).*(AMD|NVIDIA)'

function usage
{
    exit 0
}

function parse_args
{
    local _ARGS=()
    local _PCIIDS=""

    while [ "$1" ]
    do
        case "$1" in
            "-v" )
                VERBOSE=1
            ;;
            "--test")
                TEST=1
                PCI_REGEX='s/^\(..\):\(..\).\(.\).*/0x\1:0x\2:0x\3/p'
                SYSID_REGEX=''
                _PCIIDS=$(printf "f0:00.0\nf1:00.0")
            ;;
        esac
        shift
    done
    
    set -- "${args[@]}"

    start_tweak "${VERBOSE}" "${_PCIIDS}"
    exit 0
}

function start_tweak
{
    if [ -z "$2" ]; then
        #GET PCI ID FOR ALL NVIDIA/AMD GPU
        local _PCIIDS=$(lspci | sed -n -e "${PCI_REGEX}")
    else
        local _PCIIDS=$(sed -n -e "${PCI_REGEX}" <<< "$2")
    fi

    if [ -z "$_PCIIDS" ]; then
        printf "No recongnizable AMD / NVIDIA Device\r\n"
        exit 0
    fi

    while read -r _PCIID; do
        local _PCIBUS=$(parse_pciid "${_PCIID}" "lspci")

        #GET SYSTEM ID
        local _IDS=$(lspci -vnn | grep -E "${_PCIBUS}${SYSID_REGEX}" -A1 | grep -Eo '([0-9a-z]{4}:[0-9a-z]{4})')
        local _IDS=(${_IDS[@]})

        local _SYSTEM_ID=${_IDS[0]}
        local _SUBSYSTEM_ID=${_IDS[1]}

        if [ "${TEST}" = 1 ]; then
            if [ $(($DISPLAY%2)) -eq 0 ]; then
                #Test NVIDIA
                local _SYSTEM_ID="10de:1b81"
                local _SUBSYSTEM_ID="10de:119d"
            else
                #TEST AMD
                local _SYSTEM_ID="1002:67df"
                local _SUBSYSTEM_ID="1462:3417"
            fi
        fi

        if [ -z "${_SYSTEM_ID}" ] || [ -z "${_SUBSYSTEM_ID}" ]; then
            if [ "$VERBOSE" = 1 ]; then
                printf "${_PCIBUS} device not recognize\r\n"
            fi
            continue
        fi

        if [ "$VERBOSE" = 1 ]; then
            printf "Tweaking device ${_SYSTEM_ID}, ${_SUBSYSTEM_ID}\r\n"
        fi

        #Parsing JSON config
        local _DRIVER=$(jq -r ".[] | select(.system_id==\"${_SYSTEM_ID}\" and .subsystem_id==\"${_SUBSYSTEM_ID}\") | .driver" "${TWEAKCONFIG}")
        local _TWEAKS=$(jq -r ".[] | select(.system_id==\"${_SYSTEM_ID}\" and .subsystem_id==\"${_SUBSYSTEM_ID}\") | .tweaks" "${TWEAKCONFIG}")
      
        if [ -z "${_DRIVER}" ] || [ -z "${_TWEAKS}" ]; then
            if [ "$VERBOSE" = 1 ]; then
                printf "Device ${_PCIBUS} with id ${_SYSTEM_ID}, subsystem id ${_SUBSYSTEM_ID} doesn't require tweak\r\n"
            fi
            continue
        fi

 
        if [ "$VERBOSE" = 1 ]; then
            printf "Driver : ${_DRIVER}\r\n"
            printf "Tweaks commands : \r\n${_TWEAKS}\r\n"
        fi
 
        case "${_DRIVER}" in
            "amdgpu" )
                tweak_amd "${_PCIID}" "${_TWEAKS}"
                ;;
            "nvidia" )
                tweak_nvidia "${_PCIID}" "${_TWEAKS}"
                ;; 
        esac
        DISPLAY=$((DISPLAY+1))
    done <<< "${_PCIIDS}"
}

function parse_pciid
{
    local _PCIID=$1
    local _TYPE=$2
    local _PCIID_ARRAY=(${_PCIID//:/ })
    local _RESULT="${_PCIID}"

    case "${_TYPE}" in
        "xorg" )
            _RESULT=$(printf "PCI:%d:%d:%d" "${_PCIID_ARRAY[@]}")
            ;;
        "lspci" )
            _RESULT=$(printf "%02x:%02x.%0x" "${_PCIID_ARRAY[@]}")
            ;;
        "rocm" )
            _RESULT=$(printf "%x" "${_PCIID_ARRAY[0]}")
            ;;
    esac
    echo "${_RESULT}"
}

function tweak_amd
{
    local _PCIID=$1
    local _TWEAKS=$2
    local _GPUID=$(parse_pciid "${_PCIID}" "rocm")
    local _COUNT=$(echo ${_TWEAKS} | jq ". | length")
    
    for (( i=0; i<${_COUNT}; i++ ))
    do
        local _CMD=$(echo ${_TWEAKS} | jq -r ".[${i}]")
        local _REALCMD=$(sed "s/@GPUID@/${_GPUID}/g;s/@ROCSMI@/${ROCSMI//\//\\/}/g" <<< "${_CMD}")
        
        if [ "$VERBOSE" = 1 ]; then
            printf "Running : ${_REALCMD}\r\n"
        fi

        #/bin/bash -c "${_REALCMD}"
        #wait
    done
}

function tweak_nvidia
{
    local _PCIID=$1
    local _TWEAKS=$2

    local _PCIBUS=$(parse_pciid "${_PCIID}" "lspci")
    local _COUNT=$(echo ${_TWEAKS} | jq ". | length")

    #Prepare temporary tweak file for xinit
    local _TWEAK_CFG=$(mktemp /tmp/tweak-XXXXXXXX.sh)
    printf "#!/bin/bash\r\n" >> ${_TWEAK_CFG}

    for (( i=0; i<${_COUNT}; i++ ))
    do
        local _CMD=$(echo ${_TWEAKS} | jq -r ".[${i}]")
        local _REALCMD=$(sed "s/@PCI_BUS@/${_PCIBUS}/g;s/@NVSET@/${NVSET//\//\\/}/g;s/@NVSMI@/${NVSMI//\//\\/}/g" <<< "${_CMD}")
        
        printf "${_REALCMD}\r\n" >> ${_TWEAK_CFG}

        if [ "$VERBOSE" = 1 ]; then
            printf "Running : ${_REALCMD}\r\n"
        fi
    done

    start_xorg "${_PCIID}" "${_TWEAK_CFG}"
}

function start_xorg
{
    local _PCIID=$1
    local _TWEAK_CFG=$2

    local _PCIBUS_XORG=$(parse_pciid "${_PCIID}" "xorg")


    local _XORG_CFG=$(mktemp /tmp/xorg-XXXXXXXX.conf)
    sed -e s,@GPU_BUS_ID@,"${_PCIBUS_XORG}",    \
        -e s,@SET_GPU_DIR@,"${DIR}", \
        config/nvidia-xorg.conf >> "${_XORG_CFG}"

    if [ "${VERBOSE}" = 1 ]; then
        printf "xinit ${_TWEAK_CFG} --bus="${PCIID}" --run-forever --  :${DISPLAY} -once -config ${_XORG_CFG} > /dev/null 2>&1 &\r\n"
    fi

    #RUN XINIT AND TWEAK
    #xinit "${_TWEAK_CFG}" --bus="${PCIID}" --run-forever --  :"${DISPLAY}" -once -config "${_XORG_CFG}" > /dev/null 2>&1 &
    #wait

    #CLEANUP
    rm -f "${_XORG_CFG}"
    rm -f "${_TWEAK_CFG}"
}

function main
{
    parse_args "$@"
    exit 0
}

main "$@"
